Add first draft of code review guidelines (#144)
o3de/community: guides/o3de-code-review-guidelines.md

> # O3DE Code Review Guidelines

Reviewer: You have inconsistent casing in the section headings. Decide on Camel Capitalization, Sentence Capitalization etc.

Author: I can fix this, good spot

Author: (should now be fixed to follow 'Sentence capitalization')

> This document contains a set of guidelines for code reviewers and change authors when contributing to Open 3D Engine (O3DE). The intention is to help pull requests (PRs) go smoothly and help the community share best practices/lessons learned when it comes to changes and reviews.

Reviewer: ```suggestion
This document contains a set of guidelines for code reviewers and Pull Request (PR) authors when contributing to Open 3D Engine (O3DE). The intention is to help PRs go smoothly and help the community share best practices and lessons learned when it comes to changes and reviews.
```

## Useful Related Material

> - [O3DE C++ Coding Standards](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#o3de-c-coding-standards-)

Author: Are those public yet? I don't see them listed here - https://github.com/o3de/sig-core/tree/main/governance. If you can please share the link with me I'd be happy to add it here. Maybe someone from @o3de/sig-testing knows where they are?

> - [O3DE C++ Coding Standards](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#o3de-c-coding-standards-)

Reviewer 2: I'm not sure, I just assumed we would have them somewhere. Maybe add a note that Python standards are coming later? Otherwise I guess we can wait until someone else adds them?

> - [O3DE C++ Coding Standards](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#o3de-c-coding-standards-)

Author: I'd love if someone would add them 😅 But yeah afaik know nothing exists right now, maybe something for @Kadino to investigate? 🙂

- [O3DE API Ref Guidelines](https://github.com/o3de/sig-core/blob/main/governance/API-Ref-Guidelines-Update.md)

> ## Change Author Guidelines

Reviewer: ```suggestion
## Pull Request Creator Guidelines
```

Author: I quite like the term 'Change Author' and this is how Google refers to them in their code review guidelines. It's definitely not a hill I'm willing to die on but I just want to check do we think 'Pull Request Creator' is a better alternative? 🤔 Technically you could create the PR without authoring it for example 😝 I could add a little section at the top to define the term 'Change Author' more clearly perhaps?

Reviewer 2: I can think of a few cases where the person posting the pull request is not the change author. The most obvious is if a team was working together on something bigger in a branch, and the pull request is to bring that branch work to Development.

Another is if we have a pull request from a contributor stall out, sometimes it makes sense for someone else to pick up the pull request to bring it to the finish line, although in that case "Pull Request Creator" isn't fully accurate, maybe "Pull Request Maintainer"?

Author: Hmm I still do think 'Change Author' can cover that, but ideally a definition of these more unusual cases could be included for clarity, I'll update this shortly

> ## Change Author Guidelines

Author: That's a good callout, I can include a mention of that somewhere

A set of guidelines for those making changes and creating PRs (the PR author)

### Keep it short

This of course isn't always possible, there will be exceptions, but to get a PR approved and merged in a reasonable time-frame, try to keep changes as contained and atomic as possible. Small incremental progress is much safer and simpler than a single large change that contains a many modifed lines and files.

### Review your own PR

The best way to catch simple mistakes and small oversights is to read through your own PR in the GitHub UI before hitting the publish button. Viewing your own code through this slightly different lens often uncovers a lot more issues than you'd expect and can help save other reviewers time.

> ### PRs are complete

Reviewer: ```suggestion
### Ensure PRs are ready for review
```

### Respond to comments

It's good practice to acknowledge you've read the reviewer's feedback. This can be as simple as leaving a 👍 or a brief comment and then resolving the conversation. If the comment was simple and has been fixed, it's usually fine to resolve the conversation. Try to avoid resolving conversations without acknowledging or responding to them either with a reply or emoji.

### Don't sweat the small stuff

Reviewers will sometimes leave optional/suggested feedback. Not everything has to be accepted, especially with certain stylistic choices (the Coding Standards should be the ultimate arbiter here to help resolve any conflicts). It's often appropriate to move some feedback to a follow-up PR that makes the relevant updates, or they can be rolled into a future change.

### Avoid mixing functional and formatting changes

When making updates to the code, try to avoid making numerous formatting only changes while also including functional updates. The best approach is to make the functional change with the most minimal, necessary changes, and create a separate PR that exclusively includes the formatting changes.

### A picture is worth a thousand words

> If your change results in any kind of visual difference, to help reviewers quickly get context on the PR, favor including screenshots and/or videos/gifs of the change in action (before/after is also very useful). This helps frame the change and can make reviewers lives much easier to see what the effect of the change is. This goes for test output too.

Reviewer: Please emphasize that adding UX screenshots is essential for reviewing these type  of changes.

### Engage UI/UX when needed

If your PR involves user interface (UI) or user experience (UX) changes, please ensure you have read how to get help from [SIG/UI-UX](https://github.com/o3de/sig-ui-ux/blob/main/governance/UX%20Intake%20Process.md). SIG maintainers can help you connect and engage with the UI/UX team.

### Engage Docs when needed

If your PR requires changes to the O3DE documentation or tutorials, ensure you've reviewed [Issue 34](https://github.com/o3de/sig-docs-community/issues/43) (currently in review). SIG maintainers can help you connect and engage with the documentation experts.

### Nudges are okay

> If your PR has been waiting a while and no one has looked at it, it's okay to start using gentle nudges to ask for feedback. This most likely will be on Discord but it's possible to also tag people in the PR comments and politely ask them to take a look. Code owners will be automatically notified but sometimes things can be missed.

Reviewer: Do folks understand what you mean by nudge here? Consider adding links to GitHubs [@mention system](https://github.blog/2011-03-23-mention-somebody-they-re-notified/) for mentioning folks. May also want to add something about reaching out to the relevant SIG on Discord as well.

Author: I do mention 'tagging' people on the PR and Discord but I could also put the @mention system in brackets too to be super clear

Author: (now added)

> > It's inevitable people get busy or will be away too, it happens. When faced with this, try and switch to something else or look for another code owner with domain knowledge who can approve the PR.

Reviewer: Maybe this is where we add a note like 
```suggestion
> It's inevitable people get busy or will be away too, it happens. When faced with this, try and switch to something else or look for another code owner with domain knowledge who can approve the PR. You can ask for reviews in the relevant SIG channels on Discord.
```

Author: This is a good callout, I've updated the wording slightly but the sentiment remains the same.

Writing code is hard and doing so in a new environment with unfamiliar faces can be even more daunting. As reviewers we want to encourage and support new contributors. We want to thank people for their efforts and offer constructive feedback where necessary. Remember to call out the highlights of a change as well as things that need improvement.

### It's about the code

When writing review feedback focus on what the code is doing, not what the change author did. Avoid 'you' if at all possible. Instead use 'we' to demonstrate our shared responsibility and ownership of the code base or refer to the code directly.

e.g.

Avoid: _"You should use a vector here has it will lead to better cache locality"_  
Prefer: _"A vector might be a good idea to use here to get better cache locality"_

### Include the _why_

When suggesting a change (unless it's something trivial) it's helpful to include the reason _why_ such a change is being proposed. This gives context to the request and can often act as a learning opportunity for the code author and other contributors who may be reviewing or observing the PR.

### Don't be late to the party

> If a PR already has two approvals (from at least one code owner) be conscious that additional feedback will interrupt the flow of a PR. Feel free to review and if a serious issue is spotted do feel empowered to take the necessary steps (request changes) but try to avoid lots of additional comments that could be addressed in a follow-up PR.

Reviewer 2: I think late feedback is good, I'd rather get it than someone hold back on good feedback. I think instead this should be phrased along the lines of being accepting of code going in and having the changes responded to in a later follow up pull request. I even don't mind and would rather get feedback on an already merged pull request than miss out on that feedback entirely.

Otherwise, we don't really have a good mechanism for annotating suggested changes outside of pull requests.

Author: I've thought about this and do agree, I've just seen some awkward situations where a PR is blocked (request changes) for minor feedback even after have multiple approvals. This guidance is to try and avoid that situation as it can be incredibly demoralizing for contributors. I'll try and be a little more accepting of these changes saying it's okay to comment, but unless very serious, refrain from adding blocking changes and be comfortable about moving the updates to a fast-follow PR

### Keep feedback focused

> Ensure feedback is targeted at the changed lines of code and not surrounding code. It isn't the change authors responsibility to fix a problem a reviewer sees that happens to be in that area of code. If something is spotted, make an issue, and follow-up with a separate PR.

Reviewer 2: I'd rather have the reviewer first ask if the pull request author is willing to make a change and be absolutely willing to take a "Sorry out of scope for this pull request" reply. Often it's much faster to make a small change to existing code, than the effort to write a follow up ticket, spin up a new branch to do that change, generate a PR for that change, and run automated review.

Author: Again this is true it's just a question of scope, and how much is asked. I think again I can clarify this okay if it's small, or if it can be moved to a separate PR to happen right after. It isn't cool though to keep piling on suggested changes in the surrounding code when the contributor has tasks to complete and things to do outside of that focus area

### Be judicious with 'Request Changes'

Request changes is a powerful tool and shouldn't be used lightly. If an apparent bug, performance problem or architecture issue is spotted, then 'Request Changes' may be used. However, often simply not approving the PR and leaving a comment is sufficient to resolve whatever problem is detected, and does so in a softer and friendlier way. Prefer this option if you can.

### Be responsive

After reviewing a PR, keep an eye on when the change author has responded and do your best to reply to their responses promptly (ideally within a 24 hour window, though this isn't always possible). If you are delayed and don't have time to respond fully, leave a comment on the PR that you're busy and will get back to it as soon as you can.

> ### Approve if you can

Reviewer: Perhaps add a note about how we don't want Perfect to get in the way of progress. Consider moving some requests for change to a followup GitHub issues.

We may also want to link to the case where folks have not added unit tests. For example, reviewers may want to consider raising with the SIG to see if someone can take on adding unit tests in a subsequent PR. We should acknowledge that not all contributors have infinite time to cover automated testing needs.

Author: I've added a reference to the aphorism 'perfect is the enemy of good', good suggestion. To be honest I'd rather not encourage PRs without unit tests (or some kind of test coverage) that is then done later (it often isn't). Depending on the situation either we can afford _not_ to have a test, or you can't make that change without a test... maybe we need a section for that? But I'd rather leave it for now (in the doc we do already mention it being okay to move feedback/changes to a follow-up PR)

> Note: This sentiment is also lifted from the excellent Google [engineering practices](https://google.github.io/eng-practices/review/reviewer/standard.html) page

_In general, reviewers should favor approving a [PR] once it is in a state where it definitely improves the overall code health of the system being worked on, even if the [PR] isn’t perfect._

> It's inevitable all changes will not get everything right, but if what is there improves the software while keeping it maintainable and efficient, a reviewer should feel empowered to approve.

Reviewer 2: this might be a good place to remind people we can always make follow up tickets. I find it can be useful to call something out in a pull request "This could use an automated test" but with a note "and I've created a ticket to backfill that automated test, we don't need to hold up this change"

### If you can't approve, take things offline

If there is a point of discussion on a PR that cannot be quickly resolved, consider reaching out directly to the change author to talk it over (either over Discord, a separate GitHub discussion or even an RFC - sometimes a voice call is even better if possible). In our experience this leads to better outcomes for both parties and avoids potential misunderstandings. Leaving a brief update on the PR with the resolution afterwards is good practice too.

## Contributors, Reviewers and Maintainers

> Once a PR is approved by two or more reviewers, an AR run can commence. Currently only [maintainers](https://github.com/orgs/o3de/teams/maintainers) can initiate builds so please feel free to ask for someone with access to start a build. The list of PRs can be found here - [Change Requests](https://jenkins.build.o3de.org/job/O3DE/view/change-requests/). Search for the PR number to review the current status of the build. Once AR has completed and there are no failures, a PR can be merged by a maintainer.

Reviewer 2: We probably want a section to call out the details of this process a bit more clearly. There's a gotcha in our automated review process that requires pull request authors to really be on top of things: The pull request just runs on the source branch, not the merge, so pull request authors need to 1) Merge from the target branch as closely to the time they run automated review as possible, 2) Merge in the pull request as closely to the time the automated review finishes as possible, and 3) Keep an eye on the target branch build after merging, because that Jenkins job does some different things than automated review and can fail when AR passed.

Author: I'm maybe okay with this but I'm not sure this is the right document for going into all that detail. I think this should be visible for maintainers/contributors but this is really trying to focus on PR feedback (giving and receiving) as much as possible. I encourage you to create a separate doc to enumerate some of the process steps you've mentioned.

# O3DE code review guidelines

This document contains a set of guidelines for code reviewers and change authors when contributing to Open 3D Engine (O3DE). The intention is to help pull requests (PRs) go smoothly and help the community share best practices and lessons learned when it comes to changes and reviews.

> Note: The guidance here applies primarily to PRs targeting the `development` branch.

## Useful related material

- [O3DE C++ Coding Standards](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#o3de-c-coding-standards-)
- [O3DE API Ref Guidelines](https://github.com/o3de/sig-core/blob/main/governance/API-Ref-Guidelines-Update.md)

## Change author guidelines

A set of guidelines for those making changes and creating PRs (the PR author).

> Note: There may be situations where the responsibility of shepherding a PR through review falls to someone who is not the originator of the change. For the purposes of this document all such roles fall under 'Change Author' as the individual is still responsible for the work being reviewed.

### Keep it short

> This of course isn't always possible, there will be exceptions, but to get a PR approved and merged in a reasonable time-frame, try to keep changes as contained and atomic as possible. Small incremental progress is much safer and simpler than a single large change that contains a many modifed lines and files.

Reviewer: nit: maybe change "atomic" to be in simpler english. maybe "as focused and self-contained as possible".

### Respond to comments

It's good practice to acknowledge you've read the reviewer's feedback. This can be as simple as leaving a 👍 or a brief comment and then resolving the conversation. If the comment was simple and has been fixed, it's usually fine to resolve the conversation. Try to avoid resolving conversations without acknowledging or responding to them either with a reply or emoji.

### Don't sweat the small stuff

Reviewers will sometimes leave optional/suggested feedback. Not everything has to be accepted, especially with certain stylistic choices (the Coding Standards should be the ultimate arbiter here to help resolve any conflicts). It's often appropriate to move some feedback to a follow-up PR that makes the relevant updates, or they can be rolled into a future change.

### Avoid mixing functional and formatting changes

When making updates to the code, try to avoid making numerous formatting only changes while also including functional updates. The best approach is to make the functional change with the most minimal, necessary changes, and create a separate PR that exclusively includes the formatting changes.

### A picture is worth a thousand words

> If your change results in any kind of visual difference, to help reviewers quickly get context on the PR, please attach screenshots and/or videos/gifs of the change in action (including before/after is also very useful). This helps frame the change and can make reviewers lives much easier to see what the effect of the change is. This goes for test output too.

Reviewer: nit: Maybe add the example of adding screenshots of script canvas graphs as an example, the source assets are almost impossible to review but a visual overview can be very illuminating.

Author: Good suggestion, I'll include a mention of this at the end along with the bit about test output.

### Be responsive

After reviewing a PR, keep an eye on when the change author has responded and do your best to reply to their responses promptly (ideally within a 24 hour window, though this isn't always possible). If you are delayed and don't have time to respond fully, leave a comment on the PR that you're busy and will get back to it as soon as you can.

### Approve if you can

> Note: This sentiment is also lifted from the excellent Google [engineering practices](https://google.github.io/eng-practices/review/reviewer/standard.html) page

_In general, reviewers should favor approving a [PR] once it is in a state where it definitely improves the overall code health of the system being worked on, even if the [PR] isn’t perfect._

It's inevitable all changes will not get everything right, but if what is there improves the software while keeping it maintainable and efficient, a reviewer should feel empowered to approve (don't let [perfect be the enemy of good](https://en.wikipedia.org/wiki/Perfect_is_the_enemy_of_good)).

### If you can't approve, take things offline

> If there is a point of discussion on a PR that cannot be quickly resolved, consider reaching out directly to the change author to talk it over (either over Discord, a separate GitHub discussion or even an RFC - sometimes a voice call is even better if possible). In our experience this leads to better outcomes for both parties and avoids potential misunderstandings. Leaving a brief update on the PR with the resolution afterwards is good practice too.

Reviewer: nit: Maybe, if you need help from other contributors raise it in the relevant SIG channel or public meetings.
